sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java

```diff
@@ -1512,20 +1512,33 @@ public int compare(Method o1, Method o2) {
                   Sets.filter(
                       propertyNamesToGetters.keySet(),
                       input -> StringUtils.getLevenshteinDistance(entry.getKey(), input) <= 2));
+
+          PipelineOptionUnexpectedPropertyException exception;
           switch (closestMatches.size()) {
             case 0:
-              throw new IllegalArgumentException(
-                  String.format("Class %s missing a property named '%s'.", klass, entry.getKey()));
+              exception =
+                  new PipelineOptionUnexpectedPropertyException(
+                      String.format(
+                          "Class %s missing a property named '%s'.", klass, entry.getKey()));
```

Comment:
I think it can be simplified by moving all of this logic into the exception, so that here you will only need to call `throw PipelineOptionUnexpectedPropertyException.create(entry.getKey(), klass, closestMatches)` without switching.

And within the `PipelineOptionUnexpectedPropertyException` I would have something like this:
```java

PipelineOptionUnexpectedPropertyException extends IllegalArgumentException {
  ...
  private PipelineOptionUnexpectedPropertyException(option, klass, closestMatches) {
      super(createMessage(option, klass, closestMatches));
      this.optionsClass = klass;
      ...
  }
  ...
  public static PipelineOptionUnexpectedPropertyException create(...) { return new ...; }
  ...  
  public Class getPipelineOptionsClass() { return optionsClass; }
  ...
  private static String createMessage(option, klass, closestMatches) {
    String message = "Class " + klass " missing a property named '" + option"'.";
    switch (closestMatches.size()) {
       case 0: return message;
       case 1: return message + " Did you mean " + ....;
       default: return message + " Did you mean " + ...;
    }
  }
}
```